deviceplugin unit test failure

This is not 100% reproducable in my cluster.

There is 20% chance of failing in my environment.

```
$ make test WHAT=k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin



Running tests for APIVersion: v1,admissionregistration.k8s.io/v1alpha1,admissionregistration.k8s.io/v1beta1,admission.k8s.io/v1beta1,apps/v1beta1,apps/v1beta2,apps/v1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2beta1,batch/v1,batch/v1beta1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,events.k8s.io/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,policy/v1beta1,rbac.authorization.k8s.io/v1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,scheduling.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,storage.k8s.io/v1alpha1,
+++ [0115 17:59:27] Running tests without code coverage
2018/01/15 17:59:30 Starting to serve on /tmp/mock.sock
2018/01/15 17:59:30 Starting to serve on /tmp/mock.sock
E0115 17:59:30.481555   29183 endpoint.go:101] listAndWatch ended unexpectedly for device plugin mock with error rpc error: code = Unavailable desc = grpc: the connection is unavailable
panic: test timed out after 4m0s

goroutine 42 [running]:
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1145 +0xf9
created by time.goFunc
	/usr/local/go/src/time/sleep.go:170 +0x44

goroutine 1 [chan receive, 4 minutes]:
testing.(*T).Run(0xc4203bc000, 0x13342fd, 0x7, 0x13b9570, 0x481901)
	/usr/local/go/src/testing/testing.go:790 +0x2fc
testing.runTests.func1(0xc4203bc000)
	/usr/local/go/src/testing/testing.go:1004 +0x64
testing.tRunner(0xc4203bc000, 0xc4204e1de0)
	/usr/local/go/src/testing/testing.go:746 +0xd0
testing.runTests(0xc42015fb20, 0x1b9ce20, 0xb, 0xb, 0x115f4a0)
	/usr/local/go/src/testing/testing.go:1002 +0x2d8
testing.(*M).Run(0xc420397f18, 0xc4204e1f70)
	/usr/local/go/src/testing/testing.go:921 +0x111
main.main()
	k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin/_test/_testmain.go:64 +0xdb

goroutine 5 [chan receive]:
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x1ba9220)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:879 +0x9f
created by k8s.io/kubernetes/vendor/github.com/golang/glog.init.0
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:410 +0x203

goroutine 14 [chan receive, 4 minutes]:
k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin.TestRun(0xc4203bc2d0)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin/endpoint_test.go:93 +0x4af
testing.tRunner(0xc4203bc2d0, 0x13b9570)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de

goroutine 15 [IO wait, 4 minutes]:
internal/poll.runtime_pollWait(0x7f38f9f09eb0, 0x72, 0xffffffffffffffff)
	/usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420182798, 0x72, 0xc420050400, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitRead(0xc420182798, 0xffffffffffffff00, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc420182780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:335 +0x1e2
net.(*netFD).accept(0xc420182780, 0x18, 0xc4200505f0, 0x13ba3d8)
	/usr/local/go/src/net/fd_unix.go:238 +0x42
net.(*UnixListener).accept(0xc42017e360, 0xc4200506a0, 0x18, 0x7e30bd)
	/usr/local/go/src/net/unixsock_posix.go:162 +0x32
net.(*UnixListener).Accept(0xc42017e360, 0x13ba388, 0xc4200ee500, 0x1b4bfa0, 0xc42017e360)
	/usr/local/go/src/net/unixsock.go:241 +0x49
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*Server).Serve(0xc4200ee500, 0x1b4bfa0, 0xc42017e360, 0x0, 0x0)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/server.go:475 +0x196
created by k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin.(*Stub).Start
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin/device_plugin_stub.go:88 +0x15c

goroutine 50 [select]:
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*addrConn).resetTransport(0xc420264200, 0x0, 0x0)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/clientconn.go:994 +0xbd6
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*addrConn).connect.func1(0xc420264200)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/clientconn.go:695 +0x2f
created by k8s.io/kubernetes/vendor/google.golang.org/grpc.(*addrConn).connect
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/clientconn.go:694 +0x40d
FAIL	k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin	240.014s
Makefile:182: recipe for target 'test' failed
make: *** [test] Error 1

```

> @CaoShuFeng Thanks very much, can you try the follow command please:

```
$ make test WHAT=k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin GOFLAGS=-v
Running tests for APIVersion: v1,admissionregistration.k8s.io/v1alpha1,admissionregistration.k8s.io/v1beta1,admission.k8s.io/v1beta1,apps/v1beta1,apps/v1beta2,apps/v1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2beta1,batch/v1,batch/v1beta1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,events.k8s.io/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,policy/v1beta1,rbac.authorization.k8s.io/v1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,scheduling.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,storage.k8s.io/v1alpha1,
+++ [0116 09:30:03] Running tests without code coverage
=== RUN   TestNewEndpoint
2018/01/16 09:30:06 Starting to serve on /tmp/mock.sock
--- PASS: TestNewEndpoint (0.00s)
=== RUN   TestRun
2018/01/16 09:30:06 Starting to serve on /tmp/mock.sock
E0116 09:30:06.666229    6242 endpoint.go:101] listAndWatch ended unexpectedly for device plugin mock with error rpc error: code = Unavailable desc = grpc: the connection is unavailable
panic: test timed out after 4m0s

goroutine 50 [running]:
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1145 +0xf9
created by time.goFunc
	/usr/local/go/src/time/sleep.go:170 +0x44

goroutine 1 [chan receive, 4 minutes]:
testing.(*T).Run(0xc420078000, 0x13342fd, 0x7, 0x13b9570, 0x481901)
	/usr/local/go/src/testing/testing.go:790 +0x2fc
testing.runTests.func1(0xc420078000)
	/usr/local/go/src/testing/testing.go:1004 +0x64
testing.tRunner(0xc420078000, 0xc4204cdde0)
	/usr/local/go/src/testing/testing.go:746 +0xd0
testing.runTests(0xc42016fb00, 0x1b9ce20, 0xb, 0xb, 0x115f4a0)
	/usr/local/go/src/testing/testing.go:1002 +0x2d8
testing.(*M).Run(0xc420521f18, 0xc4204cdf70)
	/usr/local/go/src/testing/testing.go:921 +0x111
main.main()
	k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin/_test/_testmain.go:64 +0xdb

goroutine 5 [chan receive]:
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x1ba9220)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:879 +0x9f
created by k8s.io/kubernetes/vendor/github.com/golang/glog.init.0
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:410 +0x203

goroutine 12 [chan receive, 4 minutes]:
k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin.TestRun(0xc4200782d0)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin/endpoint_test.go:93 +0x4af
testing.tRunner(0xc4200782d0, 0x13b9570)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de

goroutine 13 [IO wait, 4 minutes]:
internal/poll.runtime_pollWait(0x7f6487a11eb0, 0x72, 0xffffffffffffffff)
	/usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42018e798, 0x72, 0xc420510400, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitRead(0xc42018e798, 0xffffffffffffff00, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc42018e780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:335 +0x1e2
net.(*netFD).accept(0xc42018e780, 0x18, 0xc4205105f0, 0x13ba3d8)
	/usr/local/go/src/net/fd_unix.go:238 +0x42
net.(*UnixListener).accept(0xc42021e960, 0xc4205106a0, 0x18, 0x7e30bd)
	/usr/local/go/src/net/unixsock_posix.go:162 +0x32
net.(*UnixListener).Accept(0xc42021e960, 0x13ba388, 0xc4200fe500, 0x1b4bfa0, 0xc42021e960)
	/usr/local/go/src/net/unixsock.go:241 +0x49
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*Server).Serve(0xc4200fe500, 0x1b4bfa0, 0xc42021e960, 0x0, 0x0)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/server.go:475 +0x196
created by k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin.(*Stub).Start
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin/device_plugin_stub.go:88 +0x15c

goroutine 15 [select]:
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*addrConn).resetTransport(0xc42054c400, 0x0, 0x0)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/clientconn.go:994 +0xbd6
k8s.io/kubernetes/vendor/google.golang.org/grpc.(*addrConn).connect.func1(0xc42054c400)
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/clientconn.go:695 +0x2f
created by k8s.io/kubernetes/vendor/google.golang.org/grpc.(*addrConn).connect
	/home/fujitsu/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/google.golang.org/grpc/clientconn.go:694 +0x40d
exit status 2
FAIL	k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin	240.020s
Makefile:182: recipe for target 'test' failed
make: *** [test] Error 1
```

--- a/pkg/kubelet/cm/deviceplugin/device_plugin_stub.go
+++ b/pkg/kubelet/cm/deviceplugin/device_plugin_stub.go
@@ -86,14 +86,11 @@ func (m *Stub) Start() error {
 	pluginapi.RegisterDevicePluginServer(m.server, m)
 
 	go m.server.Serve(sock)
-	// Wait till grpc server is ready.
-	for i := 0; i < 10; i++ {
-		services := m.server.GetServiceInfo()
-		if len(services) > 0 {
-			break
-		}
-		time.Sleep(1 * time.Second)
+	_, conn, err := dial(m.socket)
+	if err != nil {
+		return err
 	}
+	conn.Close()
 	log.Println("Starting to serve on", m.socket)
 
 	return nil
@@ -109,7 +106,8 @@ func (m *Stub) Stop() error {
 
 // Register registers the device plugin for the given resourceName with Kubelet.
 func (m *Stub) Register(kubeletEndpoint, resourceName string) error {
-	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(10*time.Second),
 		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
 			return net.DialTimeout("unix", addr, timeout)
 		}))

--- a/pkg/kubelet/cm/deviceplugin/endpoint.go
+++ b/pkg/kubelet/cm/deviceplugin/endpoint.go
@@ -186,9 +186,10 @@ func (e *endpointImpl) stop() {
 	e.clientConn.Close()
 }
 
-// dial establishes the gRPC communication with the registered device plugin.
+// dial establishes the gRPC communication with the registered device plugin. https://godoc.org/google.golang.org/grpc#Dial
 func dial(unixSocketPath string) (pluginapi.DevicePluginClient, *grpc.ClientConn, error) {
-	c, err := grpc.Dial(unixSocketPath, grpc.WithInsecure(),
+	c, err := grpc.Dial(unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(10*time.Second),
 		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
 			return net.DialTimeout("unix", addr, timeout)
 		}),

--- a/pkg/kubelet/cm/deviceplugin/manager_test.go
+++ b/pkg/kubelet/cm/deviceplugin/manager_test.go
@@ -497,9 +497,6 @@ type TestResource struct {
 
 func TestPodContainerDeviceAllocation(t *testing.T) {
 	flag.Set("alsologtostderr", fmt.Sprintf("%t", true))
-	var logLevel string
-	flag.StringVar(&logLevel, "logLevel", "4", "test")
-	flag.Lookup("v").Value.Set(logLevel)
 	res1 := TestResource{
 		resourceName:     "domain1.com/resource1",
 		resourceQuantity: *resource.NewQuantity(int64(2), resource.DecimalSI),

